package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
-            modular_pipeline_id: ID of the modular pipeline to check existence in the repository.
-        Returns:
-            Whether the given modular pipeline ID is in the repository.
-        Example:
-            >>> modular_pipelines = ModularPipelinesRepository()
-            >>> modular_pipelines.has_modular_pipeline("__root__")
-            True
-            >>> modular_pipelines.has_modular_pipeline("doesnt exist")
-            False
+            parent_node (ModularPipelineNode): The parent modular pipeline node.
+            modular_pipeline_id (str): The ID of the modular pipeline to add as a child.
+            modular_pipeline_inputs_outputs (Set[str]): A set of inputs/outputs
+                    to/from the modular pipeline
         """
-        return modular_pipeline_id in self.tree
+
+        parent_node.children.add(
+            ModularPipelineChild(
+                id=modular_pipeline_id, type=GraphNodeType.MODULAR_PIPELINE
+            )
+        )
+        for dataset in modular_pipeline_inputs_outputs:
+            if (
+                dataset not in parent_node.inputs
+                and dataset not in parent_node.outputs
+                and dataset not in self.parameters
+            ):
+                parent_node.children.add(
+                    ModularPipelineChild(id=dataset, type=GraphNodeType.DATA)
+                )
+            if dataset not in self.node_mod_pipeline_map:
+                self.node_mod_pipeline_map[dataset] = set()
+            self.node_mod_pipeline_map[dataset].add(modular_pipeline_id)
+
+    def _add_datasets_as_children(
+        self,
+        modular_pipeline: ModularPipelineNode,
+        kedro_node: Node,
+        modular_pipeline_inputs_outputs: Set[str],
+    ):
+        """Helper to add datasets (not parameters) related to task nodes as children.
+
+        Here we follow the below rule:
+        - Inputs/Outputs of a task node are added as children
+          to the modular pipeline if they are not inputs/outputs
+          of that modular pipeline and are not parameter datasets
+        """
+        hashed_io_ids = {
+            _hash_input_output(io)
+            for io in set(kedro_node.inputs) | set(kedro_node.outputs)
+        }
+        for io_id in hashed_io_ids:
+            if (
+                io_id not in modular_pipeline_inputs_outputs
+                and io_id not in self.parameters
+            ):
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=io_id, type=GraphNodeType.DATA)
+                )
+                if io_id not in self.node_mod_pipeline_map:
+                    self.node_mod_pipeline_map[io_id] = set()
+                self.node_mod_pipeline_map[io_id].add(modular_pipeline.id)
+
+    def get_node_and_modular_pipeline_mapping(
+        self, node
+    ) -> Tuple[str, Union[Set[str], None]]:
+        """Get the modular pipeline(s) to which the given node belongs."""
+        node_id = (

```

Comment:
To avoid things like this, I would've created a bit more clever `_hash` helper, which can correctly hash the dataset or the node.

Reply:
dataset or task node*   -- i tried to make this function a bit better but in this function a  'node' can be both dataset or task node. But u r right, the use of 'node' is getting a bit confusing.

Reply:
I also like the reversal here, start with the shorter one and finish with the longer one.